lucet-idl/src/witx/ast.rs

```diff
@@ -0,0 +1,161 @@
+#![allow(dead_code)]
+use std::collections::HashMap;
+use std::rc::{Rc, Weak};
+
+pub use super::parser::BuiltinType;
+
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+pub struct Id(String);
+
+impl Id {
+    pub fn new<S: AsRef<str>>(s: S) -> Self {
+        Id(s.as_ref().to_string())
+    }
+    pub fn as_str(&self) -> &str {
+        self.0.as_str()
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct Document {
+    pub definitions: Vec<Definition>,
+    pub entries: HashMap<Id, Entry>,
+}
+
+#[derive(Debug, Clone)]
+pub enum Definition {
+    Datatype(Rc<Datatype>),
+    Module(Rc<ModuleDef>),
+}
+
+#[derive(Debug, Clone)]
+pub enum Entry {
+    Datatype(Weak<Datatype>),
+    Module(Weak<ModuleDef>),
+}
+
+impl Entry {
+    pub fn kind(&self) -> &'static str {
+        match self {
+            Entry::Datatype { .. } => "datatype",
+            Entry::Module { .. } => "module",
+        }
+    }
+}
+
+#[derive(Debug, Clone)]
+pub enum DatatypeIdent {
```

Comment:
That's fair, but it seems awfully convenient during codegen to just have an `Rc` to solve the indirection rather than having to look it up elsewhere. I'm going to keep it.